### ISSUE TITLE
How about rewriting the client using Tauri?

### ISSUE THREAD

**Issue by yuezk** [yuezk/GlobalProtect-openconnect#205]
Reasons are:

- I'm more familiar with the front-end tech stack so I may be more responsive to the issues
- Could have fewer problems with the browser-based authentication
- UI could be improved

I just wanted to let you know that I don't promise it. But I want to listen to your voices.

**Comment by yuezk** [comment]
Hi, @makar2 I have a question. Is there a user-friendly way & secure way to run a process as root or call a service run as root in GUI applications?

**Comment by bagnaram** [comment]
I appreciate the work although I like the idea of a browser-agnostic approach. I personally don't use electron and am a Firefox user so I will continue to support that method. Is there a way to offload the browser and just launch the XDG default browser? Sort of what AWS-SSO and other login flows use for graphical browser based auth?

**Comment by yuezk** [comment]
@bagnaram After some research, I will not choose Electron. I could use [Tarui](https://tauri.app/) or [Neutralino.js](https://neutralino.js.org) to render the UI. The benefit is that the package size won't be large.

It has a possibility to use the default browser to do the authentication because I found some related fields in the gateway response of the GlobalProtect. I will test it when rewriting.

**Comment by piiih** [comment]
@yuezk are u still rewriting the application, is there anything that I can help you?

**Comment by yuezk** [comment]
Yes, the latest code is on the refactor branch, it works for non-browser based authentication. I didn’t provide the doc for now, you might not be able to run it.

**Comment by piiih** [comment]
@yuezk I could run the project but I'm getting `Failed to prelogin: Network failed` error, I don't know if you already finished the refactor but if you need any help to test I can do this :)

**Comment by yuezk** [comment]
@piiih Thanks for your feedback. It would help a lot before releasing the refactored client. The error message is not for external browser authentication. It could relate to the SSL certificate.

I just uploaded the latest code base to log the details of the error. The README is also updated, and you can open the DevTools and view the more detailed error logs in the Console. Would you please update to the latest code and try it again?

**Comment by yuezk** [comment]
If the detailed error is like below, you could try the solution 

**Error**
"Network Error: Tls Error: error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl/statem/extensions.c:893:: Tls Error: error:0A000152:SSL routines:final_renegotiate:unsafe legacy renegotiation disabled:ssl/statem/extensions.c:893"

**Solution**

Create a file `openssl.cnf` with the content:

```
openssl_conf = openssl_init

[openssl_init]
ssl_conf = ssl_sect

[ssl_sect]
system_default = system_default_sect

[system_default_sect]
# Use the following options if the version of openssl >= 3.0.4
Options = UnsafeLegacyServerConnect
# Use the following options if the version of openssl <= 3.0.3
# Options = UnsafeLegacyRenegotiation
```

Run the client with `OPENSSL_CONF=path/to/openssl.cnf`, for example: `OPENSSL_CONF=path/to/openssl.cnf pnpm tauri dev`

**Comment by hrk** [comment]
If I understand correctly (from their homepage), Tauri uses WebKitGTK.

saml-gp-gui already uses GTK, whereas this project's use of Qt was its advantage for users of KDE Plasma desktop.

Is it possible to keep the Qt project alive?

**Comment by yuezk** [comment]
Hi @hrk, the KDE Plasma can run the Tauri application, so why keep the Qt project? Any reason?

**Comment by hrk** [comment]
From my **personal** point of view...
99% of Linux desktops either use the Qt (KDE Plasma, others) or the GTK libraries (Gnome, Mate, XFCE, others).
When faced with the issue of having to connect to GlobalProtect/Palo Alto VPN, I found two projects: gp-saml-gui (GTK-based) and GlobalProtect-openconnect (Qt-based).
As a user of KDE Plasma, I chose the Qt-based one in order to reduce installed dependencies, active memory usage (only one framework loaded), possibly storage (libraries already present and used by the desktop).

Tauri uses GTK, so when the switch will be complete, there will be: gp-saml-gui (GTK) and GlobalProtect-openconnect-tauri (GTK).

Thanks for reading all of this :-)
